pkg/golinters/internal/diff.go

```diff
@@ -166,70 +147,88 @@ func (p *hunkChangesParser) handleAddedOnlyLines(addedLines []string) {
 	}
 
 	// add-only change merged into the last original line with possible prepending
-	p.ret = append(p.ret, Change{
-		LineRange: result.Range{
-			From: p.lastOriginalLine.originalNumber,
-			To:   p.lastOriginalLine.originalNumber,
-		},
-		Replacement: result.Replacement{
-			NewLines: append(p.replacementLinesToPrepend, append([]string{p.lastOriginalLine.data}, addedLines...)...),
-		},
-	})
+	change := Change{
+		From:     p.lastOriginalLine.originalNumber,
+		To:       p.lastOriginalLine.originalNumber,
+		NewLines: slices.Concat(p.replacementLinesToPrepend, []string{p.lastOriginalLine.data}, addedLines),
+	}
+
+	p.changes = append(p.changes, change)
+
 	p.replacementLinesToPrepend = nil
 }
 
-func (p *hunkChangesParser) parse(h *diffpkg.Hunk) []Change {
-	p.parseDiffLines(h)
+func parseDiffLines(h *diffpkg.Hunk) []diffLine {
+	lines := bytes.Split(h.Body, []byte{'\n'})
 
-	for i := 0; i < len(p.lines); {
-		line := p.lines[i]
-		if line.typ == diffLineOriginal {
-			p.handleOriginalLine(line, &i)
-			continue
-		}
+	currentOriginalLineNumber := int(h.OrigStartLine)
 
-		var deletedLines []diffLine
-		for ; i < len(p.lines) && p.lines[i].typ == diffLineDeleted; i++ {
-			deletedLines = append(deletedLines, p.lines[i])
+	var diffLines []diffLine
+
+	for i, line := range lines {
+		dl := diffLine{
+			originalNumber: currentOriginalLineNumber,
 		}
 
-		var addedLines []string
-		for ; i < len(p.lines) && p.lines[i].typ == diffLineAdded; i++ {
-			addedLines = append(addedLines, p.lines[i].data)
+		if i == len(lines)-1 && len(line) == 0 {
+			// handle last \n: don't add an empty original line
+			break
 		}
 
-		if len(deletedLines) != 0 {
-			p.handleDeletedLines(deletedLines, addedLines)
-			continue
+		lineStr := string(line)
+
+		switch {
+		case strings.HasPrefix(lineStr, "-"):
+			dl.typ = diffLineDeleted
+			dl.data = strings.TrimPrefix(lineStr, "-")
+			currentOriginalLineNumber++
+
+		case strings.HasPrefix(lineStr, "+"):
+			dl.typ = diffLineAdded
+			dl.data = strings.TrimPrefix(lineStr, "+")
+
+		default:
+			dl.typ = diffLineOriginal
+			dl.data = strings.TrimPrefix(lineStr, " ")
+			currentOriginalLineNumber++
 		}
 
-		// no deletions, only additions
-		p.handleAddedOnlyLines(addedLines)
+		diffLines = append(diffLines, dl)
 	}
 
-	if len(p.replacementLinesToPrepend) != 0 {
-		p.log.Infof("The diff contains only additions: no original or deleted lines: %#v", p.lines)
-		return nil
+	// if > 0, then the original file had a 'No newline at end of file' mark
+	if h.OrigNoNewlineAt > 0 {
+		dl := diffLine{
+			originalNumber: currentOriginalLineNumber + 1,
+			typ:            diffLineAdded,
+			data:           "",
+		}
+		diffLines = append(diffLines, dl)
 	}
 
-	return p.ret
+	return diffLines
 }
 
-func ExtractIssuesFromPatch(patch string, lintCtx *linter.Context, linterName string, formatter fmtTextFormatter) ([]result.Issue, error) {
+func ExtractDiagnosticFromPatch(pass *analysis.Pass, file *ast.File, patch string,
```

Comment:
> I wonder if they ever consider switching over to something else given that the analysis package is stable with suggested fixes.

There is no better solution because there formatters and not linters.
We have only the source file and the formatted file content, so diff is the best approach.